Philip dev

For124 opened the pull request.

For124 committed:
Merge pull request #4 from For124/Philip-dev

Philip dev

philipSchwinghammer committed:
Added Molecular Dynamics simulation. Not yet tested beyond compiling/resulting in the correct kind of data. Not yet added plotting functionality.

philipSchwinghammer committed:
Added Quantum_Basics.jl, a notebook for explaining the axioms of quantum mechanics and doing basic exercises with them.

philipSchwinghammer committed:
Fixed some bugs/mistakes regarding the bloch sphere

Luapulu committed:
Add methods.jl

Luapulu committed:
Simplify

philipSchwinghammer committed:
Added Bloch.jl file

Luapulu commented on intro.jl:
This shouldn't be in the PR. It's just a version change.

For124 replied:
Not sure I get what you mean here, aren't all changes in PR?

Luapulu replied:
You can undo the change and commit on your branch, then you won't change master when you merge.

Luapulu commented on methods.jl:
This should be in master laready. maybe rebase from master?

For124 replied:
Not sure what's going on here, I just merged it, should be fine I think

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -48,15 +48,15 @@ function Base.getindex(H::FinDiffHamiltonian{T}, i::Int, j::Int) where T
         if iy == jy
             Vij = H.potential(T((ix-1) // H.nlength), T((iy-1) // H.nwidth))
-            return (T(H.nlength) / H.length)^2 + (T(H.nwidth) / H.width)^2 + π^2/2*(1/H.length^2*H.kpointx^2 + 1/H.width^2*H.kpointy^2) + Vij
+            return (Complex(H.nlength) / H.length)^2 + (Complex(H.nwidth) / H.width)^2 + π^2/2*(1/H.length^2*H.kpointx^2 + 1/H.width^2*H.kpointy^2) + Vij
```
Rather than converting to a complex number in each branch, just add a type hint at the end of the function to convert any outputs.

```julia
function my_func(x::T, y::T, z::T)::Complex{T} where T <: Real
    # do some maths
    x * y * z
end
```

For124 replied:
I'll remember that for the future, can't be bothered to fix it here, since it won't be used anyway.

Luapulu replied:
I guess we're probably gonna delete everything in src?

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -48,15 +48,15 @@ function Base.getindex(H::FinDiffHamiltonian{T}, i::Int, j::Int) where T
     end
 
-    return zero(T)
+    return zero(Complex)
```
`zero(T)` with implicit conversion or `zero(Complex{T})` would be better. I suspect this code isn't type stable. Check with `@code_typed`.

For124 replied:
zero(T) is wrong, since T is the type of the length/width parameters - that is actually an issue with the way FinDiffHamiltonian is implemented, since the implicit assumption is that the type of the elements of the Hamiltonian is the same as that in which the length is specified, which is not generally a reasonable thing to assume. Would probably be better in general to just have the user specify which type the elements should be, and assume ComplexF64 by default?

Luapulu replied:
The idea is to output `zero(T)` but have `getindex` convert to `Complex` of that implicitly by putting a type declaration on the output. I agree that my implementation had only real components and for your Bloch case you need to change the element type to `Complex{T} where T<:Real` rather than `T<:Real`.

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -103,4 +103,96 @@ function calculate_bandstructure(V, L, W, NL, NW, path)
 end
 
+function Force_Calculator(F::Function, V::Function, NELECT::Int, NL::Int, NW::Int, L::Float64, W::Float64)
+    #We calculate the Hellman-Feynmann forces, which are the expectation value of the derivative of the Hamiltonian with respect the state. In the Born-Oppenheimer approximation we assume that the system is always in the ground state, and we have a slater determinant of the NELECT eigenstates corresponding to the lowest energy eigenvalues. We don't actually have to calculate this, since the expec ...
```
Style conventions suggest adding newlines so no line is longer than about 100 chars and adding a space between the `#` and the first character of the comment. It might be a good idea to make this into markdown text in a new cell.

For124 replied:
Will keep in mind

Luapulu replied:
I just realised this isn't even in a notebook. The character limit per line is still useful.

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -103,4 +103,96 @@ function calculate_bandstructure(V, L, W, NL, NW, path)
     return energies
 end
 
+function Force_Calculator(F::Function, V::Function, NELECT::Int, NL::Int, NW::Int, L::Float64, W::Float64)
```
julia naming convention is to make variables lower cased. `SCREAMING_CASE` tends to be used for global constants.

Luapulu replied:
There's also no requirement to make these type constraints so specific. Callable, Integer and Real would be better. Arguably, you could even remove type constraints alltogether.

Finally, function names should be lower cased without underscores when well readable and with underscores added if it helps readability.

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -103,4 +103,96 @@ function calculate_bandstructure(V, L, W, NL, NW, path)
+#module QuantumAcademy is used as a force calculator, where it solves the Schrödinger equation and then calculates the Hellman-Feynmann forces on the ions. 
+
+# Define the Molecular_Structure type
+mutable struct Molecular_Structure
```
This doesn't need to be mutable if you're changing entries in the vectors. immutable structs are preferred when possible.

Luapulu commented on src/QuantumAcademy.jl:
Have a look at the julia naming conventions.

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -103,4 +103,96 @@ function calculate_bandstructure(V, L, W, NL, NW, path)
+            total_dVdy[i] += dVdy
+        end
+        # The force is the negative gradient of the potential
+        return [-total_dVdx, -total_dVdy]
```
This will allocate a new vector on every call. Better would be to return a StaticArray form StaticArrays.jl or just a tuple.

Luapulu commented on src/QuantumAcademy.jl:
```diff
@@ -103,4 +103,96 @@ function calculate_bandstructure(V, L, W, NL, NW, path)
+mutable struct Molecular_Structure
+    length::Float64
+    width::Float64
+    positions::Vector{Tuple{Float64, Float64}}
```
two separate vectors is probaly better than one vector of tuples. In general, column based storage tends to be more memory efficient, leading to fewer cache misses and therefore faster code. As always, benchmarking is the only way to know if it actually matters.

Luapulu commented on test/runtests.jl:
```diff
@@ -72,3 +72,28 @@ using LinearAlgebra
+    @test Structure.length isa Float64
+    @test Structure.width isa Float64
+    Structuredeltat = QuantumAcademy.propagate(Structure, 0.001, NL, NW)
+    @show sum(force_field(Structure)(x/NL*L,y/NW*W) for x in 1:NL for y in 1:NW)
```
I assume this is debugging code you probably want to remove.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+
+
+# ╔═╡ 29bfbd20-b307-4ae6-a2b2-ad4c556c4479
+Spinstate = Float64[1,0]#We initialize a vector with two entries. The state [1,0] will correspond to spin up, and [0,1] to spin down. 
```
should be something like `state = [1, 0]`. Check naming conventions and leave spaces after commas.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+An observeable is a physical quantity that we can measure. Examples include the energy, the momentum, the position and the spin of a particle. Each such observeable can be represented by a hermitian matrix, which acts on the state vector. In our example of looking at spin states, one example would be the spin of the particle along the z axis, i.e. whether the spin is pointing up (1) or down (-1). ...
+
+# ╔═╡ 9bade9d9-2301-4f36-8809-344873f8630c
+σ_z = Float64[1 0; 0 -1]
```
These could just be ints. Alos, `σz` is better than `σ_z`. This is not python.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+σ_x = Float64[0 1; 1 0]
+
+# ╔═╡ e8cd222b-97f7-42d6-834b-c80621a76a2f
+σ_y = ComplexF64[0 -im; im 0]
```
type descriptions aren't needed. Julia will do the right thing automatically.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+
+# ╔═╡ c8602d92-88cd-4c98-b8c3-47575247e01a
+function probability(vec1, vec2)#Write a function which takes two vectors, and returns the squared norm of the scalar product. 
+	return norm(dot(vec1,vec2))^2 
```
Isn't there a norm2 function? Or `abs2` would do it.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+md"**Once the measurement has been performed, the state of the system will be the eigenstate belonging to the eigenvalue that was returned by the measurement**"
+
+# ╔═╡ 7331fca3-7713-4e97-b442-14266d98b9c4
+function perform_measurement_Z(state)#This function takes a state and an operator, and simulates performing a measurement of the spin in z-direction. It returns the state after the measurement was performed. 
```
How about `zmeasure`? A bit more Julian.

Luapulu replied:
Ideally, you'd write random functions like this:

```julia
function zmeasure(rng::AbstractRNG, state)
    # use the local rng, rather than the global one
    r = rand(rng)
end

# add a convenience method
zmeasure(state) = zmeasure(Random.default_rng(), state)
```

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+
+# ╔═╡ 9a790930-9dc4-47a0-bf62-1ad65f68887d
+function expectation_value_Z(state)#Calculates the expectation value for measuring the spin along the z-axis for a given state
+	return dot(state, σ_z * state)
```
`dot(state, sigma_z, state)` is better.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+Show that a global complex phase does not change the outcome of measurements!"
+
+# ╔═╡ 592fafe8-ec65-486c-a541-f38da30d2dd2
+if probability(eigvecs(σ_x)[:,2], [0,1]) == probability(eigvecs(σ_x)[:,2]*cis(3), [0,1])
```
SInce this is inside a notebook, you could also just return the string, which will display nicely, rather than printing in console.

Luapulu commented on Quantum_Basics.jl:
Cool, looks nice!

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+ - How much the state points in the z-direction"
+
+# ╔═╡ f6b42b7d-c614-4c95-a105-d7dced988825
+function parametrize_Bloch_sphere(state)
```
How about `parameterize_bloch`? Fixed the typo and shortened the name a bit.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+
+plotted_points = zeros(Float64, (3,length(time_range)))
+
+for i in 1:length(time_range)
```
Maybe split this up into multiple cells? I'm not sure, but I find multiple cells often makes code look cleaner, even if it is longer. The exception for me is plotting code.

A more elegant solution might be to use `zip` and `eachcol`. So, 

```Julia
pnts = zeros(Float64, (3, times))
for (t, col) in zip(times, eachcol(pnts))
    col .= to_bloch(propagator(t) * initial_state)
end
```

I've renamed a few things to be a bit shorter. Maybe a function `propagate(time, state)` would be good, too. This way you can avoid allocations. You could also use StaticArrays. Performance doesn't matter at this stage, but it's good to get in the habit of avoiding allocations when you don't need to allocate.

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+# ╔═╡ f0201489-9e3c-462e-8978-3c47ed7fa569
+function Hamiltonian_t(t)
+	return [sin(t) cos(t); cos(t) -2sin(t)]
+end
```
How about
```Julia
Ht(t) = [sin(t) cos(t); cos(t) -2sin(t)]
```

Luapulu commented on Quantum_Basics.jl:
```diff
@@ -0,0 +1,1888 @@
+# ╔═╡ 72d32760-3c54-406b-b59a-1d4a8f087488
+function Generic_Time_Evolution(time)#Write the function which returns the time evolution operator U(t) for a generic Hamiltonian. Assume ħ=1.
+	return cis(Hamiltonian_t(time)*time)
+end
```
This is wrong isn't it? You'd need to integrate the Hamiltonian up to time `t` wouldn't you?

For124 replied:
Yeah true, I'll fix that later

Luapulu replied:
I wonder how you'd best do the integration. Maybe [FFTW.jl](https://github.com/JuliaMath/FFTW.jl) would be best, since you can do the Fourier transform once, and then use it to evaluate the integral at different points.

If that's too much work for you, consider Gauss or Clenshaw-Curtis weights and nodes:
- [FastGaussQuadrature.jl](https://github.com/JuliaApproximation/FastGaussQuadrature.jl) for Gauss quadrature.
- [FastTransforms.jl](https://juliaapproximation.github.io/FastTransforms.jl/dev/#FastTransforms.clenshawcurtisnodes) for Clenshaw Curtis.

If all else fails, I guess we're left with trapezoid or Simpon's rule or something like that.

Luapulu reviewed (commented):
Looking good. I think there's a mistake in the time dependent solution. As I said, watch out for stylistic conventions, allocations and too constrained type bounds.

philipSchwinghammer committed:
Improved Bloch.jl by adding plotting of charge densities
Added Molecular_Dynamics.jl, almost finished; Can run and plot MD that looks almost correct, needs more work on boundary conditions, energy scaling, numerical accuracy

For124 committed:
Merge branch 'master' into Philip-dev

Luapulu commented:
I think you should've waited on merging.
